src/connect4/core.rs

```diff
@@ -257,64 +278,103 @@ impl Board {
-                    if !space_used && run_len >= 4 {
+                //If token not for team in cell, end of search in target direction
+                if val != 0 && val != team {
+                    dir_active = false;
+                //If 0 or 1 spaces in target direction used, either add to run_len and/or potential_run depending on if cell is empty or matches team
+                } else if dir_spaces_used <=  1 {
+                    //If you have a contiguous run of 4 with no spaces, immediately return because a winning run has been found!
+                    if run_len >= 4 && dir_spaces_used == 0 {
                         return 4i32;
+                    } else if val == team {
+                        run_len += 1;   
+                    } else {
+                        dir_spaces_used +=1;
                     }
-                } else if val == 0 && !space_used {
-                    space_used = true;
+                    potential_len += 1;
+                //If more than one space in target direction used, only add to potential length for non-enemy cells
                 } else {
-                    dir_active = false;
+                    potential_len += 1;
                 }
             }
             i += 1;
         }
-        if space_used {
-            std::cmp::min(run_len, 3)
-        //Todo: Handle case where "run" is blocked on both ends
+        //If the potential of the run is not 4 or more, return 0 because it is not a viable run
+        if potential_len < 4 {
+            return 0i32;
+        //Otherwise, return the minimum of the run_len and 4 (if no spaces) or 3 (if one space used)
         } else {
-            std::cmp::min(run_len, 4)
+            if dir_spaces_used > 0 {
+                std::cmp::min(run_len, 3)
+            } else {
+                std::cmp::min(run_len, 4)
+            }
         }
     }
 
     //Method to return an array of runs from a start location for a given team, where array[i] returns the number of runs
     //of length i-1. Accounts for all eight directions, but may have false duplicates (e.g. a run BAAAB will return have two
     //runs of length 3 for team A even though technically its the same run)
-    fn get_runs_from_point(&self, start: GridPosition, team: i32) -> [i32;4] {
+    pub fn get_runs_from_point(&self, start: GridPosition, team: i32) -> [i32;4] {
         let mut output = [0i32;4];
         let directions = vec![(1, 0), (1, 1), (0, 1), (-1, 1)];
         for dir in directions {
-            output[(self.get_run_in_direction(start, GridPosition::new(dir.0, dir.1), team)-1) as usize] += 1;
-            output[(self.get_run_in_direction(start, GridPosition::new(-1*dir.0, -1*dir.1), team)-1) as usize] += 1;
+            let a = self.get_run_in_direction(start, GridPosition::new(dir.0, dir.1), team)-1;
+            let b = self.get_run_in_direction(start, GridPosition::new(-1*dir.0, -1*dir.1), team)-1;
+            if a >= 0 {
```

Comment:
Had to make a separate check because `get_run_in_direction` can return values from 0-4, but only values corresponding to runs of length 1-4 are held.